### ISSUE TITLE
Fabric 2.0 support

### ISSUE THREAD

**Issue by swapnildahiphale** [APGGroeiFabriek/PIVT#53]
Do you plan to upgrade your sources for Hyperledger Fabric 2.0?

**Comment by LudoNieuw** [comment]
Hi there,

We have it on our backlog, but it’s not a high priority for us at the
moment, being busy with other stuff. If you have some ideas or want to make
a start, please do, as we can all benefit.

Best,

Ludo

Op do 28 mei 2020 om 19:04 schreef Swapnil Dahiphale <
notifications@github.com>

**Comment by mr-rooftop** [comment]
To support the 2.0 CC lifecycle, for `commit` and `invoke` the `tlsRootCertFiles` of all other peers who will participate in the CC are needed.

In the Argo workflows, only the orderer tls secrets and the secrets of the current peer are mounted.
What would be the cleanest way give each peer access to the `tlsRootCertFiles` of all other peers?

Mount all tls secrets in all namespaces/clusters?
Copy some certs around as it is done in the `README` for the Genesis Block creation?
Something else?

I'm currently speaking about the `cross-cluster setup`.

Some pointers would be greatly appreciated, as the combination of Helm + Argo + K8s + HF is not the easiest to understand.

**Comment by mr-rooftop** [comment]
Ok figured out how to do it.
I post it here for documentation purposes.

As explained here https://github.com/APGGroeiFabriek/PIVT#preperation, there are already secrets for external Peer Msp secrets:

> In cluster-one you will notice two external MSP secrets:
> 
> `hlf-peer--atlantis--external-msp`
> `hlf-peer--nevergreen--external-msp`
> 
> These are created because of ExternalPeerOrgs in crypto-config.yaml and required for channel creation.
> 
> In cluster-one and cluster-three you will notice external orderer TLS CA secret:
> 
> `hlf-orderer--pivt-external-tlsca`
> 
> These are created because of ExternalOrdererOrgs in crypto-config.yaml. In cluster-one it's not used but in cluster-three required to communicate with external orderer.

Basically I used the same logic to get the peer `tlsRootCerts` that are needed for the new HF 2.0 chaincode lifecycle into the argo workflow container.

---
In `cluster-one/` there are already entries for `ExternalPeerOrgs` in the `samples/cross-cluster-raft-tls/cluster-one/crypto-config.yaml`. 
Now extend the `crypto-config.yaml` of `cluster-two` and `cluster-three` so that they have as `ExternalPeerOrgs` the Orgs of the other clusters.
E.g. extendsion in `samples/cross-cluster-raft-tls/cluster-one/crypto-config.yaml`:
```
ExternalPeerOrgs:
  - Name: Karga
    Domain: aptalkarga.tr
  - Name: Nevergreen
    Domain: nevergreen.nl
```
Note that Karga is Peer Org of cluster-one, Nevergreen is Peer Org of cluster-three, and Atlantis is already in `PeerOrgs:` of the `crypto-config` as it isn't an external Peer Org.

---
The external k8s secrets get created via the helm chart `hlf-kube/templates/peerorg-external-msp-secret.yaml`.
In the current implementation the helm chart puts `config.yaml`, `admincert.pem` and `cacert.pem` in the k8s secret. What we need however for the chaincode lifecycle is `tlscacert.pem`.
That's why one needs to extend the helm chart with the additional file, the resulting `hlf-kube/templates/peerorg-external-msp-secret.yaml` is this:
```
{{- range $i, $org := .Values.ExternalPeerOrgs }}

apiVersion: v1
kind: Secret
type: Opaque
metadata:
  name: hlf-peer--{{ $org.Name | lower }}--external-msp
  labels:
    orgName: {{ $org.Name }}
data:
  config.yaml: |- 
      {{ $.Files.Get (printf "crypto-config/peerOrganizations/%s/msp/config.yaml" $org.Domain ) | b64enc }}
  admincert.pem: |- 
      {{ $.Files.Get (printf "crypto-config/peerOrganizations/%s/msp/admincerts/Admin@%s-cert.pem" $org.Domain $org.Domain ) | b64enc }}
  cacert.pem: |- 
      {{ $.Files.Get (printf "crypto-config/peerOrganizations/%s/msp/cacerts/ca.%s-cert.pem" $org.Domain $org.Domain ) | b64enc }}
  tlscacert.pem: |- 
      {{ $.Files.Get (printf "crypto-config/peerOrganizations/%s/msp/tlscacerts/tlsca.%s-cert.pem" $org.Domain $org.Domain ) | b64enc }}
---
{{- end }} {{- /* ExternalPeerOrgs */ -}}
```

---
Thanks to the above steps, we have the tls certs of every peer in every namespace/cluster as K8s secrets, and we can access them in the `chaincode-flow/templates/chaincodes-workflow.yaml`! :heart: 

You can iterate over them with the `.Values.ExternalPeerOrgs` attribute which is provided when you call the workflow as described in the `README` with `samples/cross-cluster-raft-tls/cluster-one/crypto-config.yaml`:
```
{{- range $i, $org := .Values.ExternalPeerOrgs }}
```

With such `range`, you need to mount the secrets as volumes:
```
{{- range $i, $org := .Values.ExternalPeerOrgs }}

  - name: peer-{{ $org.Name | lower }}--external-msp
    secret:
      secretName: hlf-peer--{{ $org.Name | lower }}--external-msp
      
{{- end }} {{- /* Orgs */ -}}{{""}}
```
you need to mount the volumes into the leaf templates of `commit` and `invoke`:
```
{{- range $i, $org := $.Values.ExternalPeerOrgs }}
    - mountPath: /hlf_config/peer-{{ $org.Name | lower }}-tlsca
      name: peer-{{ $org.Name | lower }}--external-msp
{{- end }} {{- /* Orgs */ -}}{{""}}
```
and in the `peer lifecycle` calls of `commit` and `invoke` you need to insert the `tlsRootCerts` into the specific `peer lifecycle` command:
```
{{- range $i, $org := $.Values.ExternalPeerOrgs }} \
--peerAddresses peer0.{{ $org.Domain | lower }}:7051 \
--tlsRootCertFiles /hlf_config/peer-{{ $org.Name | lower }}-tlsca/tlscacert.pem \
{{- end }} {{- /* Orgs */ -}}{{""}}
```
Furthermore, you cannot use the docker container from @raftAtGit `raft/hl-fabric-tools:1.4.3`, because it doesn't support the HF 2.0 chaincode lifecycle. 
Instead, use the official image `hyperledger/fabric-tools:2.0`.
Beware that the new container run on Alpine Linux, which (as far as I understand) does not support `bash`, but only `sh` (this means that slightly different commands are available in container shell).
So the script field of your leaf template should start like this:
```
script:
image: hyperledger/fabric-tools:2.0
command: [sh]
```

---
Last but not least, you need to copy some more certs around as described in the `README`.

In https://github.com/APGGroeiFabriek/PIVT#preperation is written:

> Lets copy them (the target empty placeholder directories are already created by init.sh script based on ExternalPeerOrgsand ExternalOrdererOrgs):
> 
> `# run in one`
> `cp -r ../fabric-kube-two/hlf-kube/crypto-config/ordererOrganizations/pivt.nl/msp/* hlf-kube/crypto-config/ordererOrganizations/pivt.nl/msp/`
> `cp -r ../fabric-kube-two/hlf-kube/crypto-config/peerOrganizations/atlantis.com/msp/* hlf-kube/crypto-config/peerOrganizations/atlantis.com/msp/`
> `cp -r ../fabric-kube-three/hlf-kube/crypto-config/peerOrganizations/nevergreen.nl/msp/* hlf-kube/crypto-config/peerOrganizations/nevergreen.nl/msp/`
> `cp ../fabric-kube-two/hlf-kube/crypto-config/ordererOrganizations/pivt.nl/orderers/orderer0.pivt.nl/tls/server.crt hlf-kube/crypto-config/ordererOrganizations/pivt.nl/orderers/orderer0.pivt.nl/tls/`

We need to copy also the certs of the externalPeerOrgs to the `cypto-config/` folders of every other peer. In the quote above this is already done for `cluster-one` in line 2 and 3.
For the other two clusters just replace the correct Orgs like this for `cluster-two`:
```
cp -r ../fabric-kube/hlf-kube/crypto-config/peerOrganizations/aptalkarga.tr/msp/* hlf-kube/crypto-config/peerOrganizations/aptalkarga.tr/msp/
cp -r ../fabric-kube-three/hlf-kube/crypto-config/peerOrganizations/nevergreen.nl/msp/* hlf-kube/crypto-config/peerOrganizations/nevergreen.nl/msp/
```

---
For the new lifecycle, first all orgs need to install and approve the CC, then one Org needs to commit it, and after that all Orgs can invoke the CC.

Because of that you need to adjust the values in `chaincode-flow/values.yaml` accordingly.
Probably you need to call the chaincode argo workflow twice with different values for the first cycle (install approve) before commit and the invocation after commit.

Also beware of the `--init-required` required flag in `approve` and `commit`. 
If this flag got set, the first time a CC gets invoked you need to provide the `--isInit` in the `peer chaincode invoke` command.

---
Once again, amazing repo, great work @LudoNieuw and @raftAtGit, 42 thanks!
You guys even take a look at issues, unbelievably awesome :pray: :love_letter:

**Comment by raftAtGit** [comment]
very nice @mr-rooftop 👍  

2 questions:
1. are you sure those are `tls` certs but not `msp` certs? I'm not much familiar Fabric 2.0 chaincode lifecycle, but in general MSP certs are used to sign something.
2. Any ideas for how to make this flow declarative?

**Comment by mr-rooftop** [comment]
Good point about the msp certs, I made a new helm template so that only the tlsca certs get used :+1: 

What do you mean with "make this flow declarative"? 
Isn't it already declarative?

**Comment by raftAtGit** [comment]
> Isn't it already declarative?

raft/hl-fabric-tools:1.4.3 `peer checkinstalled/checkinstantiated` commands will not work with Fabric 2 chaincode lifecycle if I'm not mistaken?

**Comment by mr-rooftop** [comment]
yes, but you can use the new commands like `queryinstalled` (https://hyperledger-fabric.readthedocs.io/en/release-2.0/commands/peerlifecycle.html#peer-lifecycle-chaincode-queryinstalled) in the `hyperledger/fabric-tools:2.0` image.

BTW, having a hard time getting golang CC to run (newbie to golang).
In your chaincode-workflow, you simply installed the CC in the $GOPATH.
However, that does not work in my setup, always this error:
```
Error: error getting chaincode bytes: failed to calculate dependencies: incomplete package: github.com/hyperledger/fabric/core/chaincode/shim
```
Do you know if I need to install the go dependencies with something like 
```
go get github.com/hyperledger/fabric/core/chaincode/shim
``` 
?

I tried a lot (https://www.edureka.co/community/29076/hyperledger-cannot-compile-chaincode-cannot-package-error, 
https://stackoverflow.com/questions/39960039/cannot-compile-chaincode, 
https://stackoverflow.com/questions/32758235/how-to-get-all-dependency-files-for-a-program-using-golang),
but nothing seems to work.

**Comment by raftAtGit** [comment]
dont know much about Go lang myself either..

about `queryinstalled` command, possibly it prints a list of installed chaincodes, like `peer chaincode list --installed` in Fabric 1.4.x and requires text parsing to process the result. 

I wanted to avoid that error-prone, ugly, fragile text parsing that's why `raft/hl-fabric-tools:1.4.3` is implemented. but unfortunately it didnt go into Fabric 1.4. code base..

**Comment by mr-rooftop** [comment]
> error-prone, ugly, fragile text parsing
I feel like everything with automation and hyperledger comes to some extent with that.

> dont know much about Go lang myself either..
Ok thx, unfortunate.

**Comment by sillysachin** [comment]
@mr-rooftop  - Can i reach you on email or linkedin - http://linkedin.com/in/sachinshekharr/ ?

**Comment by abhisamant7** [comment]
> ```
> go get github.com/hyperledger/fabric/core/chaincode/shim
> ```

Issue has been resolve on [here](https://github.com/APGGroeiFabriek/PIVT/issues/52#issuecomment-683291298).

**Comment by Marek00Malik** [comment]
@mr-rooftop could you please explain how did you change the channel-workflow template to use the peer commands?
